encode type values in binary and move type context into zng (#2826)
brimdata/zed: docs/formats/zng.md

|0x19|<ncolumns>|<name1><typeval><name2><typeval>...|
-----------------------------------------------------
```
where `<ncolumns>` is the number of columns in the record encoded as a `uvarint`,
`<name1>` etc. are the field names encoded as in the
record typedef, and each `<typeval>` is a recursive encoding of a type value.

#### 4.2 Array Type Value

An array type value has the form:
```
----------------
|0x20|<typeval>|
----------------
```
where `<typeval>` is a recursive encoding of a type value.

#### 4.3 Set Type Value

An set type value has the form:
```
----------------
|0x21|<typeval>|
----------------
```
where `<typeval>` is a recursive encoding of a type value.

#### 4.4 Union Type Value

A union type value has the form:
```
-------------------------------------
|0x22|<ntypes>|<typeval><typeval>...|
-------------------------------------
```
where `<ntypes>` is the number of types in the union encoded as a `uvarint`
and each `<typeval>` is a recursive definition of a type value.

#### 4.5 Enum Type Value

An enum type value has the form:
```
--------------------------------
|0x23|<nelem>|<name1><name2>...|
--------------------------------
```

> where `<nelem>` and each symbol name are encoded as in an enum typedef.

Reviewer: ```suggestion
where `<nelem>` and each symbol name is encoded as in an enum typedef.
```

| `bstring`  | 17 | variable | UTF-8 byte sequence defined by ZSON           |
| `ip`       | 18 | 4 or 16  | 4 or 16 bytes of IP address                    |
| `net`      | 19 | 8 or 32  | 8 or 32 bytes of IP prefix and subnet mask     |
| `type`     | 20 | variable | type value byte sequence [as defined below](#4-type-values) |
| `error`    | 21 | variable | UTF-8 byte sequence of string of error message |
| `null`     | 22 |    0     | No value, always represents an undefined value |

> TBD: For decimal, ZSON assumes decimal128 and there's not a way to specify
> different widths.  This should be okay since a 128 will hold the others.

## 4. Type Values

> As the ZSON data model support first-class types and because the ZNG designs

Reviewer: ```suggestion
As the ZSON data model support first-class types and because the ZNG design
```

> goals require that values need not be translated between type contexts, type values

Reviewer: ```suggestion
goals require that values need not be translated when used with other type contexts, type values
```

Author: I massage this a bit more...

must be encoded in a fashion that is independent of the type context.
Thus, a serialized type value encodes the entire type in a canonical form
according to the recursive definition in this section.

The type value of a primitive type (include type `type`) is its primitive ID,
serialized as a single byte.

The type value of a complex type is serialized recursively according to the

> complex type it represents as described below.
> 
> 
> #### 4.1 Record Type Value

Reviewer: ```suggestion
complex type it represents as described below.

#### 4.1 Record Type Value
```

A record type value has the form:
```
-----------------------------------------------------
|0x19|<ncolumns>|<name1><typeval><name2><typeval>...|
-----------------------------------------------------
```
where `<ncolumns>` is the number of columns in the record encoded as a `uvarint`,

> `<name1>`, `<name2>`, etc. are the field names encoded as in the
> record typedef and each `<typeval>` is a recursive encoding of a type value.

Reviewer: ```suggestion
`<name1>` etc. are the field names encoded as in the
record typedef, and each `<typeval>` is a recursive encoding of a type value.
```

#### 4.2 Array Type Value

An array type value has the form:
```
----------------
|0x20|<typeval>|
----------------
```
where `<typeval>` is a recursive encoding of a type value.

#### 4.3 Set Type Value

An set type value has the form:
```
----------------
|0x21|<typeval>|
----------------
```

> where `<typeval>` is a recursive encoding of a type value.
> 
> 
> #### 4.4 Union Type Value

Reviewer: ```suggestion
where `<typeval>` is a recursive encoding of a type value.

#### 4.4 Union Type Value
```

A union type value has the form:
```
-------------------------------------
|0x22|<ntypes>|<typeval><typeval>...|
-------------------------------------
```
where `<ntypes>` is the number of types in the union encoded as a `uvarint`
and each `<typeval>` is a recursive definition of a type value.

#### 4.5 Enum Type Value

An enum type value has the form:
```
--------------------------------
|0x23|<nelem>|<name1><name2>...|
--------------------------------
```
where `<nelem>` and each symbol name are encoded as in an enum typedef.

#### 4.6 Map Type Value

A map type value has the form:
```
----------------------------
|0x23|<key-type>|<val-type>|
----------------------------
```

> where `<key-type>` and `<val-type>` are recursive encodings of type values.
> 
> 
> #### 4.7 Alias Type Value

Reviewer: ```suggestion
where `<key-type>` and `<val-type>` are recursive encodings of type values.

#### 4.7 Alias Type Value
```
